formats/protobuf/src/main/java/io/cloudevents/protobuf/ProtoSerializer.java

```diff
@@ -0,0 +1,261 @@
+         * @param protoBuilder The {@link CloudEvent.Builder} to set attributes on.
+         * @param spec The spec to set attributes for.
+         */
+        public ProtoContextWriter(CloudEvent.Builder protoBuilder, SpecVersion spec) {
+            this.protoBuilder = protoBuilder;
+            this.requiredAttributeNumberMap = versionToAttrs.get(spec);
+            if (this.requiredAttributeNumberMap == null) {
+                throw CloudEventRWException.newInvalidSpecVersion(spec.toString());
+            }
+        }
+
+        /**
+         * Checks if this attribute is a required field, and if so, set it. This is slightly
+         * wonky because in the protobuf representation for cloud events, required attributes are
+         * fields, whereas optional attributes and extensions are stored in a map.
+         *
+         * @param name The name of the attribute
+         * @param value The value to set
+         * @return true if name was a required field that was set, false otherwise.
+         */
+        private boolean setRequiredField(String name, Object value) {
+            FieldDescriptor fieldDescriptor = this.requiredAttributeNumberMap.get(name);
+            if (fieldDescriptor == null) {
+                return false;
+            }
+            this.protoBuilder.setField(fieldDescriptor, value);
+            return true;
+        }
+
+        @Override
+        public CloudEventContextWriter withContextAttribute(String name, String value)
+            throws CloudEventRWException {
+            if (!setRequiredField(name, value)) {
+                this.protoBuilder.putAttributes(name,
+                    CloudEventAttributeValue.newBuilder().setCeString(value).build());
+            }
+            return this;
+        }
+
+        @Override
+        public CloudEventContextWriter withContextAttribute(String name, URI value)
+            throws CloudEventRWException {
+            if (!setRequiredField(name, value.toString())) {
+                // This is a bit of a hack. The java SDK doesn't differentiate between absolute and reference for its URIs, unlike the protobuf representation.
+                // So here, if we are certain this is an absolute URI, then put it in the URI type, otherwise the UriRef type.
+                final CloudEventAttributeValue.Builder builder = CloudEventAttributeValue.newBuilder();
+                if (value.isAbsolute()) {
+                    this.protoBuilder.putAttributes(name, builder.setCeUri(value.toString()).build());
+                } else {
+                    this.protoBuilder.putAttributes(name, builder.setCeUriRef(value.toString()).build());
+                }
+            }
+            return this;
+        }
+
+        @Override
+        public CloudEventContextWriter withContextAttribute(String name, OffsetDateTime value)
+            throws CloudEventRWException {
+            Instant instant = value.toInstant();
```

Comment:
I might be mistaken here, but I think you need to make sure `value` is adjusted at `UTC` before converting to `Timestamp`. You might need to use this method: https://docs.oracle.com/javase/8/docs/api/java/time/OffsetDateTime.html#withOffsetSameInstant-java.time.ZoneOffset-

Reply:
The getEpochSeconds() call returns seconds relative to UTC. I added a test to make sure this is the case.